### ISSUE TITLE
Issue with "python3 get-pip.py --user" with python 3.6.7

### ISSUE THREAD

**Issue by ayeerr** [pypa/get-pip#43]
Hi 

I am new python and pip and encountered some problems while installing pip.

Any suggestions ?

Thanks
ayeerr

$ python3 --version
Python 3.6.7
$ python3 get-pip.py --user
Traceback (most recent call last):
  File "get-pip.py", line 21361, in <module>
    main()
  File "get-pip.py", line 197, in main
    bootstrap(tmpdir=tmpdir)
  File "get-pip.py", line 82, in bootstrap
    import pip._internal
  File "/tmp/tmpm7edordz/pip.zip/pip/_internal/__init__.py", line 40, in <module>
  File "/tmp/tmpm7edordz/pip.zip/pip/_internal/cli/autocompletion.py", line 8, in <module>
  File "/tmp/tmpm7edordz/pip.zip/pip/_internal/cli/main_parser.py", line 8, in <module>
  File "/tmp/tmpm7edordz/pip.zip/pip/_internal/cli/cmdoptions.py", line 14, in <module>
ModuleNotFoundError: No module named 'distutils.util'

**Comment by yangmeishu** [comment]
the same issue 
os:wsl ubuntu 1804
version:3.6.7

**Comment by yangmeishu** [comment]
resolve : sudo apt-get install python3-distutils

**Comment by nirorman** [comment]
the same issue
os:wsl ubuntu 1804
version:3.7

**Comment by 0xmjk** [comment]
@nirorman I had the same issue with 3.7 as well.

I hacked around it like this:
```sh
apt-get download python3-distutils
dpkg-deb -x python3-distutils /
curl https://bootstrap.pypa.io/get-pip.py | python3.7
```

What it does is that it downloads the package with apt-get without installing. Then dpkg-deb extracts it outside of apt/dpkg processes to / (whilst ignoring dependencies).

Why it works is that the `python3-distutils` package (at least in its current state) provides for both Python versions - 3.6 and 3.7. However, it also pulls all python3 packages, whereas I wanted only python3.7 packages. 

Hope it helps!

**Comment by shankarprabhakaran9** [comment]
pi@raspberrypi:~ $ sudo python3 get-pip.py
Traceback (most recent call last):
  File "get-pip.py", line 21492, in <module>
    main()
  File "get-pip.py", line 197, in main
    bootstrap(tmpdir=tmpdir)
  File "get-pip.py", line 82, in bootstrap
    import pip._internal
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/__init__.py", line 40, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/cli/autocompletion.py", line 8, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/cli/main_parser.py", line 12, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/commands/__init__.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/commands/completion.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/cli/base_command.py", line 20, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/download.py", line 26, in <module>
  File "<frozen importlib._bootstrap>", line 1032, in _handle_fromlist
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_vendor/six.py", line 92, in __get__
    result = self._resolve()
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_vendor/six.py", line 115, in _resolve
    return _import_module(self.mod)
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_vendor/six.py", line 82, in _import_module
    __import__(name)
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 724, in exec_module
  File "<frozen importlib._bootstrap_external>", line 857, in get_code
  File "<frozen importlib._bootstrap_external>", line 525, in _compile_bytecode
ValueError: bad marshal data (invalid reference)

**Comment by shankarprabhakaran9** [comment]
i am getting error if i do pip install python3

pi@raspberrypi:~ $ sudo python3 get-pip.py
Traceback (most recent call last):
  File "get-pip.py", line 21492, in <module>
    main()
  File "get-pip.py", line 197, in main
    bootstrap(tmpdir=tmpdir)
  File "get-pip.py", line 82, in bootstrap
    import pip._internal
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/__init__.py", line 40, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/cli/autocompletion.py", line 8, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/cli/main_parser.py", line 12, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/commands/__init__.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/commands/completion.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/cli/base_command.py", line 20, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_internal/download.py", line 26, in <module>
  File "<frozen importlib._bootstrap>", line 1032, in _handle_fromlist
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_vendor/six.py", line 92, in __get__
    result = self._resolve()
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_vendor/six.py", line 115, in _resolve
    return _import_module(self.mod)
  File "/tmp/tmps2sh7lpy/pip.zip/pip/_vendor/six.py", line 82, in _import_module
    __import__(name)
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 724, in exec_module
  File "<frozen importlib._bootstrap_external>", line 857, in get_code
  File "<frozen importlib._bootstrap_external>", line 525, in _compile_bytecode
ValueError: bad marshal data (invalid reference)

**Comment by jonseymour** [comment]
> @nirorman I had the same issue with 3.7 as well.
> 
> I hacked around it like this:
> 
> ```shell
> apt-get download python3-distutils
> dpkg-deb -x python3-distutils /
> curl https://bootstrap.pypa.io/get-pip.py | python3.7
> ```
> 
> What it does is that it downloads the package with apt-get without installing. Then dpkg-deb extracts it outside of apt/dpkg processes to / (whilst ignoring dependencies).
> 
> Why it works is that the `python3-distutils` package (at least in its current state) provides for both Python versions - 3.6 and 3.7. However, it also pulls all python3 packages, whereas I wanted only python3.7 packages.
> 
> Hope it helps!

This mostly worked for me, thank you.

I did have to make sure I ran the extract in a directory that the _apt user could read (I chose /tmp) and I needed to specify the full file name of the python3-distutils.deb, but other than that it worked well, so thank you.

**Comment by adamdry** [comment]
Thanks, I also had to do this `sudo apt install python3-testresources` and run the script with sudo `sudo python3 get-pip.py`

**Comment by pradyunsg** [comment]
You shouldn't sudo python get-pip.py, that'll mangle files managed by your operating system installer.

**Comment by pradyunsg** [comment]
For clarity for folks dealing with this, this occurs because someone over on the Debian side decided to actively exclude distutils from the default python.

We've flagged this to the Debian folks before and it's not an issue that pip can do much about.

**Comment by Law101** [comment]
> resolve : sudo apt-get install python3-distutils

Thank you!

**Comment by moxy-bc** [comment]
I have installed python3-distutils, but now I get this:

pi@rpi3b:~/opencv/projects $ pip freeze
Traceback (most recent call last):
  File "/usr/local/bin/pip", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/dist-packages/pip/_internal/cli/main.py", line 73, in main
    command = create_command(cmd_name, isolated=("--isolated" in cmd_args))
  File "/usr/local/lib/python3.7/dist-packages/pip/_internal/commands/__init__.py", line 96, in create_command
    module = importlib.import_module(module_path)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/lib/python3.7/dist-packages/pip/_internal/commands/freeze.py", line 8, in <module>
    from pip._internal.cache import WheelCache
  File "/usr/local/lib/python3.7/dist-packages/pip/_internal/cache.py", line 8, in <module>
    import json
  File "/usr/lib/python3.7/json/__init__.py", line 106, in <module>
    from .decoder import JSONDecoder, JSONDecodeError
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 724, in exec_module
  File "<frozen importlib._bootstrap_external>", line 857, in get_code
  File "<frozen importlib._bootstrap_external>", line 525, in _compile_bytecode
ValueError: bad marshal data (unknown type code)

**Comment by chrahunt** [comment]
Hi @moxy-bc, it looks like your Python installation may be the problem since the actual exception is raised outside of pip's code and does not depend on any input that pip provides. I would try searching and posting on stackoverflow, that should get your issue better exposure.

**Comment by pradyunsg** [comment]
I'm going to go ahead and summarize this issue one last time.

For users facing this issue on Ubuntu (or anything else Debian based), this occurs because Debian decided that `distutils` is not a "core Python" package, despite being a part of the Python standard library. All of Python Packaging tooling is written to assume that `distutils` is available.

The way to fix this would be to run:

```
$ sudo apt-get install python3-distutils
```

**Comment by pradyunsg** [comment]
For folks facing issue with the marshal bug, please file a new issue here.

**Comment by engelmav** [comment]
Downgrading distutils worked for me.

**Comment by rachmadaniHaryono** [comment]
i get here from because this is first search result for my error

this happen on my python3.7 installation but not my main python3 after i upgrade my ubuntu to 20.04

```shell
➜  ~ python3 --version
Python 3.6.10
➜  ~ python3.7 -m pip --version
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/q/.local/lib/python3.7/site-packages/pip/__main__.py", line 16, in <module>
    from pip._internal.cli.main import main as _main  # isort:skip # noqa
  File "/home/q/.local/lib/python3.7/site-packages/pip/_internal/cli/main.py", line 10, in <module>
    from pip._internal.cli.autocompletion import autocomplete
  File "/home/q/.local/lib/python3.7/site-packages/pip/_internal/cli/autocompletion.py", line 9, in <module>
    from pip._internal.cli.main_parser import create_main_parser
  File "/home/q/.local/lib/python3.7/site-packages/pip/_internal/cli/main_parser.py", line 7, in <module>
    from pip._internal.cli import cmdoptions
  File "/home/q/.local/lib/python3.7/site-packages/pip/_internal/cli/cmdoptions.py", line 19, in <module>
    from distutils.util import strtobool
ModuleNotFoundError: No module named 'distutils.util'
```
- solution 1: install python3-distutils

it is already installed and reinstall it change nothing

- solution 2: download python3-distutils deb, install deb, download get-pip & run python3 get-pip

it failed when run i python3 on script

```shell
➜  ~ curl https://bootstrap.pypa.io/get-pip.py | python3.7
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1824k  100 1824k    0     0   264k      0  0:00:06  0:00:06 --:--:--  358k
Traceback (most recent call last):
  File "<stdin>", line 23470, in <module>
  File "<stdin>", line 198, in main
  File "<stdin>", line 82, in bootstrap
  File "/tmp/tmpct3zaj9_/pip.zip/pip/_internal/cli/main.py", line 10, in <module>
  File "/tmp/tmpct3zaj9_/pip.zip/pip/_internal/cli/autocompletion.py", line 9, in <module>
  File "/tmp/tmpct3zaj9_/pip.zip/pip/_internal/cli/main_parser.py", line 7, in <module>
  File "/tmp/tmpct3zaj9_/pip.zip/pip/_internal/cli/cmdoptions.py", line 19, in <module>
ModuleNotFoundError: No module named 'distutils.util'
```
what working for me:

- add deadsnake ppa https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa
- install python3.7

**Comment by sarthak73** [comment]
I am having the same issue and I tried above all things none of them worked.
I am on **Ubuntu 20.04** and have *python3.6, python3.7, python3.8*
but `python3.6 -m pip install <package-name>`  
Gives me error  
- `ModuleNotFoundError: No module named 'distutils.util'`  
- `ModuleNotFoundError: No module named 'apt_pkg'`

noob here please help me out, python3.6 is necessary can't use other versions.  
Thanks in advance.

**Comment by HemilTheRebel** [comment]
TLDR; Removing virtualenv directory and acitivating a new virtual env worked

I had this issue. One fine day, pip decided to want to make me die cause it was working perfectly just the day before. None of the above solutions worked for me. Then I went old school and deleted the venv and activated a new one and installed the dependencies of my project and it worked perfectly

**Comment by imSunnyk** [comment]
> Thanks, I also had to do this `sudo apt install python3-testresources` and run the script with sudo `sudo python3 get-pip.py`

Thank you ....it worked for me

**Comment by upss1988** [comment]
The same issue on PyCharm and Ubuntu 20.

This is the fix: `sudo apt-get install python3-distutils`

**Comment by eskp** [comment]
This makes it impossible to install pip3 as non root user as installing of python3-distutils requires root privs.

**Comment by ahaghgoo** [comment]
what would be the command to install on QNAP?
